File: Graphics_Hello_World/cpp/ppm.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

bool outputFile(std::string const &fileName,
  std::string const &header,
  std::vector<std::string> const &pixels) {
    std::ofstream output;
    try {
      output.open(fileName);
      output << header.c_str();
      for (auto rgb : pixels) {
        output << rgb.c_str();
      }
      return true;
    } catch (...) {
      std::cout << "Any error occured writing image" << std::endl;
    }
    return false;
  }

std::string generateHeader(int width, int height) {
  std::string header = "P3\n" + std::to_string(width) + " " + std::to_string(height) + "\n255\n";
  return header;
}

std::vector<std::string> generatePixels(int width, int height) {
  std::vector<std::string> pixels;
  pixels.reserve(width * height);
  std::string rgb;
  for (int j = height-1; j >= 0; j--) {
    for (int i = 0; i < width; i++) {
      float r = float(i) / float(width-1);
      float g = float(j) / float(height-1);
      float b = 0.25;
      int ir = int(255.999 * r);
      int ig = int(255.999 * g);
      int ib = int(255.999 * b);
      rgb = std::to_string(ir) + " " + std::to_string(ig) + " " + std::to_string(ib) + "\n";
      pixels.emplace_back(rgb);
    }
  }
  return pixels;
}

int main() {
  int width = 256;
  int height = 256;
  auto header = generateHeader(width, height);
  auto pixels = generatePixels(width, height);
  outputFile("cppimage.ppm", header, pixels);
}

File: Graphics_Hello_World/nim/ppm.nim

proc outputFile(fileName: string, header: string, pixels: seq[string]): bool =
  var output = open(fileName, fmWrite)
  try:
    output.writeLine(header)
    for p in pixels:
      output.writeLine(p)
    return true
  except:
    echo "Any error occured writing image"
  return false

proc generateHeader(width: int, height: int): string =
  var header: string = "P3\n" & $width & " " & $height & "\n255"
  return header

proc generatePixels(width: int, height: int): seq[string] =
  var pixels = newSeq[string]()
  var rgb: string

  for j in countdown(height-1, 0):
    for i in 0..<width:
      let r: float = float(i) / float(width-1)
      let g: float = float(j) / float(height-1)
      let b: float = 0.25
      let ir: int = int(255.999 * r)
      let ig: int = int(255.999 * g)
      let ib: int = int(255.999 * b)
      rgb = $ir & " " & $ig & " " & $ib
      pixels.add(rgb)

  return pixels

# main
let x = 256
let y = 256
let header = generateHeader(x, y)
let pixels = generatePixels(x, y)
let image = outputFile("nimimage.ppm", header, pixels)
if image == false:
  echo "Error writing image file"
